File: android/app/src/main/java/com/alexivanov/snapsell/ui/review/ReviewScreen.kt
package com.alexivanov.snapsell.ui.review

import android.graphics.Rect
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.R
import com.alexivanov.snapsell.ui.common.ErrorText
import com.alexivanov.snapsell.ui.common.GhostButton
import com.alexivanov.snapsell.ui.common.LoadingBox
import com.alexivanov.snapsell.ui.common.LucideIcon
import com.alexivanov.snapsell.ui.common.PrimaryButton
import com.alexivanov.snapsell.ui.common.SecondaryButton
import com.alexivanov.snapsell.ui.common.SnapBottomBar
import com.alexivanov.snapsell.ui.common.SnapTopBar
import com.alexivanov.snapsell.ui.common.Spinner
import com.alexivanov.snapsell.ui.theme.SnapType
import com.alexivanov.snapsell.ui.theme.snapColors
import com.alexivanov.snapsell.vision.Segment
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/** Where the photo lands inside its 4:3 box (fit, centred). */
private data class Fit(val scale: Float, val dx: Float, val dy: Float) {
    fun toBitmapX(x: Float) = (x - dx) / scale
    fun toBitmapY(y: Float) = (y - dy) / scale
}

private fun fitInto(bitmapW: Int, bitmapH: Int, boxW: Float, boxH: Float): Fit {
    val scale = min(boxW / bitmapW, boxH / bitmapH)
    return Fit(scale, (boxW - bitmapW * scale) / 2f, (boxH - bitmapH * scale) / 2f)
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun ReviewScreen(
    container: AppContainer,
    onRetake: () -> Unit,
    onConfirm: (List<String>) -> Unit,
    onBack: () -> Unit,
) {
    val vm: ReviewViewModel = viewModel { ReviewViewModel(container) }
    val state by vm.state.collectAsStateWithLifecycle()
    val c = snapColors

    // Box mode: the dragged rectangle stays as a draft until "Place box" commits it.
    var dragStart by remember { mutableStateOf<Offset?>(null) }
    var dragEnd by remember { mutableStateOf<Offset?>(null) }
    var draft by remember { mutableStateOf<Rect?>(null) }
    fun clearDraft() { dragStart = null; dragEnd = null; draft = null }

    Scaffold(
        containerColor = c.bg,
        topBar = { SnapTopBar(title = "Review items", onBack = onBack) },
        // Pinned so the primary action is never below the fold on a tall photo.
        bottomBar = {
            if (state.bitmap != null && !state.committing) {
                SnapBottomBar {
                    val n = state.selectedCount
                    PrimaryButton(
                        label = if (n == 0) "Select at least one object" else "Identify $n item${if (n == 1) "" else "s"}",
                        onClick = { vm.identify(onConfirm) },
                        enabled = n > 0 && !state.segmenting,
                        modifier = Modifier.weight(1f),
                    )
                }
            }
        },
    ) { padding ->
        val bitmap = state.bitmap
        if (bitmap == null) {
            Column(Modifier.padding(padding).padding(24.dp)) {
                ErrorText(state.error ?: "No photo.")
                PrimaryButton("Take a photo", onClick = onRetake, modifier = Modifier.padding(top = 16.dp).fillMaxWidth())
            }
            return@Scaffold
        }
        if (state.committing) {
            LoadingBox(modifier = Modifier.padding(padding), label = "Cutting out ${state.selectedCount} object(s)")
            return@Scaffold
        }

        Column(Modifier.padding(padding).fillMaxSize().verticalScroll(rememberScrollState())) {
            if (state.outlinerUnavailable) {
                Column(Modifier.padding(start = 16.dp, end = 16.dp, top = 14.dp).fillMaxWidth().border(2.dp, c.accent).padding(14.dp)) {
                    Text("On-device outlining unavailable", style = SnapType.bodySmall.copy(fontWeight = FontWeight.ExtraBold), color = c.accent700)
                    Text(
                        "Your device can't run the object outliner. Draw a box around each thing you want to sell.",
                        style = SnapType.bodySmall,
                        color = c.text,
                        modifier = Modifier.padding(top = 4.dp),
                    )
                }
            }

            val imageBitmap = remember(bitmap) { bitmap.asImageBitmap() }
            Box(
                Modifier
                    .padding(horizontal = 16.dp, vertical = 14.dp)
                    .fillMaxWidth()
                    .aspectRatio(4f / 3f)
                    .border(1.dp, c.divider),
            ) {
                PhotoCanvas(
                    imageBitmap = imageBitmap,
                    bitmapW = bitmap.width,
                    bitmapH = bitmap.height,
                    segments = state.segments,
                    selected = state.selected,
                    manualMode = state.manualMode,
                    dragStart = dragStart,
                    dragEnd = dragEnd,
                    draft = draft,
                    onTap = { x, y -> vm.tapAt(x, y) },
                    onDragStart = { dragStart = it; dragEnd = it; draft = null },
                    onDrag = { dragEnd = it },
                    onDragEnd = { rect -> draft = rect; dragStart = null; dragEnd = null },
                    onDragCancel = { dragStart = null; dragEnd = null },
                )
                if (state.manualMode && draft == null && dragStart == null) {
                    Box(Modifier.matchParentSize(), contentAlignment = Alignment.Center) {
                        Text("Drag a box around the object", style = SnapType.body.copy(fontWeight = FontWeight.ExtraBold), color = Color.White)
                    }
                }
                if (state.segmenting) {
                    Box(Modifier.matchParentSize().background(Color.Black.copy(alpha = 0.35f)), contentAlignment = Alignment.Center) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Spinner(color = Color.White)
                            Text("  Finding objects", style = SnapType.bodySmall, color = Color.White)
                        }
                    }
                }
            }
            Text(
                "${state.segments.size} object${if (state.segments.size == 1) "" else "s"} found",
                style = SnapType.fieldLabel.copy(fontSize = 11.sp),
                color = c.text.copy(alpha = 0.5f),
                modifier = Modifier.padding(horizontal = 16.dp),
            )

            Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
                ErrorText(state.error)

                if (state.segments.isNotEmpty()) {
                    // Chip pairs: a 44dp "Object N" button and a 40dp retake button divided by a 2dp rule.
                    FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        state.segments.forEach { seg ->
                            val on = seg.index in state.selected
                            val fg = if (on) c.bg else c.text
                            Row(
                                Modifier
                                    .then(if (on) Modifier.background(c.accent).border(2.dp, c.accent) else Modifier.border(2.dp, c.divider))
                                    .height(44.dp),
                                verticalAlignment = Alignment.CenterVertically,
                            ) {
                                Box(Modifier.fillMaxHeight().clickable { vm.toggle(seg.index) }.padding(horizontal = 14.dp), contentAlignment = Alignment.Center) {
                                    Text("Object ${seg.index + 1}", style = SnapType.chipLabelSelected.copy(fontSize = 13.sp), color = fg)
                                }
                                Box(Modifier.width(2.dp).fillMaxHeight().background(if (on) c.bg.copy(alpha = 0.5f) else c.divider))
                                Box(Modifier.size(40.dp).clickable { vm.retake(seg.index) { onRetake() } }, contentAlignment = Alignment.Center) {
                                    LucideIcon(R.drawable.ic_lucide_refresh_cw, "Retake object ${seg.index + 1}", size = 16.dp, tint = fg)
                                }
                            }
                        }
                    }
                }

                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    if (state.manualMode) {
                        SecondaryButton(
                            label = "Place box",
                            onClick = { draft?.let { vm.addManualRect(it) }; clearDraft() },
                            enabled = draft != null,
                            minHeight = 48.dp,
                            modifier = Modifier.weight(1f),
                        )
                    } else {
                        SecondaryButton("Add item", onClick = { clearDraft(); vm.setManualMode(true) }, minHeight = 48.dp, modifier = Modifier.weight(1f))
                    }
                    SecondaryButton("Retake photo", onClick = onRetake, minHeight = 48.dp, modifier = Modifier.weight(1f))
                }
                if (state.manualMode && state.segments.isNotEmpty()) {
                    GhostButton("Cancel box", onClick = { clearDraft(); vm.setManualMode(false) }, minHeight = 36.dp)
                }
            }
        }
    }
}

/**
 * The photo, fit into its box, with bounding-box outlines: 2.5dp solid accent
 * when selected, 2dp dashed white @75% when not, each with a 22dp numbered
 * badge at its top-left corner. In box mode a scrim covers the photo and the
 * drag rectangle is drawn dashed white.
 */
@Composable
private fun PhotoCanvas(
    imageBitmap: androidx.compose.ui.graphics.ImageBitmap,
    bitmapW: Int,
    bitmapH: Int,
    segments: List<Segment>,
    selected: Set<Int>,
    manualMode: Boolean,
    dragStart: Offset?,
    dragEnd: Offset?,
    draft: Rect?,
    onTap: (Int, Int) -> Unit,
    onDragStart: (Offset) -> Unit,
    onDrag: (Offset) -> Unit,
    onDragEnd: (Rect) -> Unit,
    onDragCancel: () -> Unit,
) {
    val c = snapColors
    val accent = c.accent
    val ink = c.text
    Canvas(
        Modifier
            .fillMaxSize()
            .background(Color(0xFF111010))
            .pointerInput(manualMode) {
                val fit = fitInto(bitmapW, bitmapH, size.width.toFloat(), size.height.toFloat())
                if (manualMode) {
                    var s: Offset? = null
                    var e: Offset? = null
                    detectDragGestures(
                        onDragStart = { s = it; e = it; onDragStart(it) },
                        onDrag = { change, _ -> e = change.position; onDrag(change.position) },
                        onDragCancel = { onDragCancel() },
                        onDragEnd = {
                            val a = s
                            val b = e
                            if (a != null && b != null) {
                                onDragEnd(
                                    Rect(
                                        fit.toBitmapX(a.x).toInt(), fit.toBitmapY(a.y).toInt(),
                                        fit.toBitmapX(b.x).toInt(), fit.toBitmapY(b.y).toInt(),
                                    ).apply { sort() },
                                )
                            } else {
                                onDragCancel()
                            }
                        },
                    )
                } else {
                    detectTapGestures { pos -> onTap(fit.toBitmapX(pos.x).toInt(), fit.toBitmapY(pos.y).toInt()) }
                }
            },
    ) {
        val fit = fitInto(bitmapW, bitmapH, size.width, size.height)
        drawImage(
            imageBitmap,
            dstOffset = IntOffset(fit.dx.toInt(), fit.dy.toInt()),
            dstSize = IntSize((bitmapW * fit.scale).toInt(), (bitmapH * fit.scale).toInt()),
        )
        val dash = PathEffect.dashPathEffect(floatArrayOf(8.dp.toPx(), 5.dp.toPx()))
        val badge = 22.dp.toPx()
        val textPaint = android.graphics.Paint().apply {
            textSize = 11.sp.toPx()
            isAntiAlias = true
            isFakeBoldText = true
            textAlign = android.graphics.Paint.Align.CENTER
        }

        fun rectOf(seg: Segment): Pair<Offset, Size> = Offset(fit.dx + seg.startX * fit.scale, fit.dy + seg.startY * fit.scale) to
            Size(seg.width * fit.scale, seg.height * fit.scale)

        if (manualMode) {
            drawRect(Color.Black.copy(alpha = 0.5f))
        }
        segments.forEach { seg ->
            val on = seg.index in selected
            val (tl, sz) = rectOf(seg)
            if (on) {
                drawRect(accent, tl, sz, style = Stroke(2.5.dp.toPx()))
            } else {
                drawRect(Color.White.copy(alpha = 0.75f), tl, sz, style = Stroke(2.dp.toPx(), pathEffect = dash))
            }
            // Numbered badge, -2dp offset from the corner.
            val bx = tl.x - 2.dp.toPx()
            val by = tl.y - 2.dp.toPx()
            drawRect(if (on) accent else Color.White, Offset(bx, by), Size(badge, badge))
            textPaint.color = if (on) android.graphics.Color.WHITE else ink.toArgbInt()
            drawContext.canvas.nativeCanvas.drawText("${seg.index + 1}", bx + badge / 2, by + badge / 2 - (textPaint.descent() + textPaint.ascent()) / 2, textPaint)
        }
        // Live drag or the placed draft, both dashed white.
        val live = if (dragStart != null && dragEnd != null) {
            Offset(min(dragStart.x, dragEnd.x), min(dragStart.y, dragEnd.y)) to Size(abs(dragEnd.x - dragStart.x), abs(dragEnd.y - dragStart.y))
        } else if (draft != null) {
            Offset(fit.dx + draft.left * fit.scale, fit.dy + draft.top * fit.scale) to Size(max(1, draft.width()) * fit.scale, max(1, draft.height()) * fit.scale)
        } else {
            null
        }
        live?.let { (tl, sz) -> drawRect(Color.White, tl, sz, style = Stroke(2.dp.toPx(), pathEffect = dash)) }
    }
}

private fun Color.toArgbInt(): Int = this.toArgb()

File: ios/SnapSell/UI/Screens/Review/ReviewScreen.swift
import SwiftUI

/// Where the photo lands inside its box (fit, centred).
private struct Fit {
    var scale: CGFloat, dx: CGFloat, dy: CGFloat
    func toImage(_ p: CGPoint) -> CGPoint { CGPoint(x: (p.x - dx) / scale, y: (p.y - dy) / scale) }
    func toView(_ r: CGRect) -> CGRect { CGRect(x: dx + r.minX * scale, y: dy + r.minY * scale, width: r.width * scale, height: r.height * scale) }
    static func into(_ image: CGSize, box: CGSize) -> Fit {
        let s = min(box.width / image.width, box.height / image.height)
        return Fit(scale: s, dx: (box.width - image.width * s) / 2, dy: (box.height - image.height * s) / 2)
    }
}

struct ReviewScreen: View {
    @Environment(AppContainer.self) private var app
    @Environment(Router.self) private var router
    @Environment(\.snap) private var c
    @State private var vm: ReviewViewModel?
    // Box mode: the dragged rectangle stays as a draft until "Place box" commits it.
    @State private var dragRect: CGRect?
    @State private var draft: CGRect?

    var body: some View {
        Group {
            if let vm { content(vm) } else { Color.clear }
        }
        .onAppear { if vm == nil { vm = ReviewViewModel(app: app) } }
        .snapScreen()
    }

    private func retake() { router.replaceTop(with: .capture) }

    @ViewBuilder
    private func content(_ vm: ReviewViewModel) -> some View {
        @Bindable var vm = vm
        VStack(spacing: 0) {
            SnapTopBar("Review items", onBack: { router.pop() })
            if let image = vm.image {
                if vm.committing {
                    LoadingBox("Cutting out \(vm.selectedCount) object(s)").padding(16)
                    Spacer()
                } else {
                    ScrollView {
                        VStack(alignment: .leading, spacing: 0) {
                            if vm.outlinerUnavailable {
                                VStack(alignment: .leading, spacing: 2) {
                                    Text("On-device outlining unavailable").font(.custom("Archivo-ExtraBold", size: 13)).foregroundStyle(c.accent700)
                                    Text("This device can't run the object outliner. Draw a box around each thing you want to sell.").snap(SnapType.bodySmall).foregroundStyle(c.text)
                                }
                                .padding(14).frame(maxWidth: .infinity, alignment: .leading)
                                .overlay(Rectangle().stroke(c.accent, lineWidth: SnapMetrics.rule))
                                .padding(.horizontal, 16).padding(.top, 14)
                            }
                            photoBox(vm, image)
                            Text("\(vm.segments.count) object\(vm.segments.count == 1 ? "" : "s") found")
                                .font(.custom("Archivo-SemiBold", size: 11)).foregroundStyle(c.text.opacity(0.5)).padding(.horizontal, 16)
                            VStack(alignment: .leading, spacing: 12) {
                                if let e = vm.error { ErrorText(e) }
                                if !vm.segments.isEmpty { chips(vm) }
                                HStack(spacing: 8) {
                                    if vm.manualMode {
                                        SnapButton("Place box", kind: .secondary, enabled: draft != nil, fullWidth: false) {
                                            if let d = draft { vm.addManualRect(d) }
                                            draft = nil; dragRect = nil
                                        }
                                    } else {
                                        SnapButton("Add item", kind: .secondary) { draft = nil; dragRect = nil; vm.manualMode = true }
                                    }
                                    SnapButton("Retake photo", kind: .secondary) { retake() }
                                }
                                if vm.manualMode, !vm.segments.isEmpty {
                                    SnapButton("Cancel box", kind: .ghost, fullWidth: false) { draft = nil; dragRect = nil; vm.manualMode = false }
                                }
                            }
                            .padding(16)
                        }
                    }
                    bottomBar(vm)
                }
            } else {
                VStack(alignment: .leading, spacing: 16) {
                    ErrorText(vm.error ?? "No photo.")
                    SnapButton("Take a photo") { retake() }
                }
                .padding(24)
                Spacer()
            }
        }
    }

    private func bottomBar(_ vm: ReviewViewModel) -> some View {
        let n = vm.selectedCount
        return VStack(spacing: 0) {
            Rule()
            SnapButton(n == 0 ? "Select at least one object" : "Identify \(n) item\(n == 1 ? "" : "s")", enabled: n > 0 && !vm.segmenting) {
                vm.identify { ids in router.toConfirm(ids) }
            }
            .padding(16)
        }
        .background(c.bg)
    }

    private func photoBox(_ vm: ReviewViewModel, _ image: UIImage) -> some View {
        let ratio = image.size.width / image.size.height
        return GeometryReader { geo in
            let w = geo.size.width
            let h = min(w / ratio, 560)
            let fit = Fit.into(image.size, box: CGSize(width: w, height: h))
            ZStack(alignment: .topLeading) {
                Color(hex: 0x111010)
                Image(uiImage: image).resizable().frame(width: image.size.width * fit.scale, height: image.size.height * fit.scale)
                    .offset(x: fit.dx, y: fit.dy)
                if vm.manualMode { Color.black.opacity(0.5) }
                ForEach(vm.segments) { seg in
                    let on = vm.selected.contains(seg.index)
                    let r = fit.toView(seg.bounds)
                    Rectangle().stroke(on ? c.accent : .white.opacity(0.75), style: StrokeStyle(lineWidth: on ? 2.5 : 2, dash: on ? [] : [8, 5]))
                        .frame(width: r.width, height: r.height).offset(x: r.minX, y: r.minY)
                    Text("\(seg.index + 1)").font(.custom("Archivo-ExtraBold", size: 11))
                        .foregroundStyle(on ? c.bg : c.text)
                        .frame(width: 22, height: 22).background(on ? c.accent : .white)
                        .offset(x: r.minX, y: r.minY)
                }
                if let d = dragRect ?? draft {
                    let r = fit.toView(d)
                    Rectangle().stroke(.white, style: StrokeStyle(lineWidth: 2, dash: [8, 5]))
                        .frame(width: r.width, height: r.height).offset(x: r.minX, y: r.minY)
                }
                if vm.manualMode, draft == nil, dragRect == nil {
                    Text("Drag a box around the object").font(.custom("Archivo-ExtraBold", size: 14)).foregroundStyle(.white)
                        .frame(width: w, height: h)
                }
                if vm.segmenting {
                    HStack(spacing: 8) { ProgressView().tint(.white); Text("Finding objects").snap(SnapType.bodySmall).foregroundStyle(.white) }
                        .frame(width: w, height: h).background(Color.black.opacity(0.35))
                }
            }
            .frame(width: w, height: h)
            .clipped()
            .contentShape(Rectangle())
            .overlay(Rectangle().stroke(c.divider, lineWidth: 1))
            .gesture(gesture(vm, fit))
        }
        .aspectRatio(max(ratio, 1 / (560 / UIScreen.main.bounds.width)), contentMode: .fit)
        .padding(.horizontal, 16).padding(.vertical, 14)
    }

    private func gesture(_ vm: ReviewViewModel, _ fit: Fit) -> some Gesture {
        DragGesture(minimumDistance: 0)
            .onChanged { g in
                guard vm.manualMode else { return }
                let a = fit.toImage(g.startLocation), b = fit.toImage(g.location)
                dragRect = CGRect(x: min(a.x, b.x), y: min(a.y, b.y), width: abs(a.x - b.x), height: abs(a.y - b.y))
            }
            .onEnded { g in
                if vm.manualMode {
                    if let r = dragRect, r.width > 4, r.height > 4 { draft = r }
                    dragRect = nil
                } else if hypot(g.translation.width, g.translation.height) < 10 {
                    vm.tap(at: fit.toImage(g.location))
                }
            }
    }

    private func chips(_ vm: ReviewViewModel) -> some View {
        FlowLayout(spacing: 8) {
            ForEach(vm.segments) { seg in
                let on = vm.selected.contains(seg.index)
                let fg = on ? c.bg : c.text
                HStack(spacing: 0) {
                    Button { vm.toggle(seg.index) } label: {
                        Text("Object \(seg.index + 1)").font(.custom("Archivo-SemiBold", size: 13)).foregroundStyle(fg)
                            .padding(.horizontal, 14).frame(height: 44)
                    }
                    .buttonStyle(.plain)
                    Rectangle().fill(on ? c.bg.opacity(0.5) : c.divider).frame(width: 2, height: 44)
                    Button { vm.retake(seg.index) { retake() } } label: {
                        Image(systemName: "arrow.clockwise").font(.system(size: 15, weight: .bold)).foregroundStyle(fg).frame(width: 40, height: 44)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Retake object \(seg.index + 1)")
                }
                .background(on ? c.accent : .clear)
                .overlay(Rectangle().stroke(on ? c.accent : c.divider, lineWidth: SnapMetrics.rule))
            }
        }
    }
}

/// Minimal wrapping HStack.
struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let width = proposal.width ?? .infinity
        var x: CGFloat = 0, y: CGFloat = 0, rowH: CGFloat = 0
        for s in subviews {
            let sz = s.sizeThatFits(.unspecified)
            if x > 0, x + sz.width > width { x = 0; y += rowH + spacing; rowH = 0 }
            x += sz.width + spacing
            rowH = max(rowH, sz.height)
        }
        return CGSize(width: width == .infinity ? x : width, height: y + rowH)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        var x: CGFloat = 0, y: CGFloat = 0, rowH: CGFloat = 0
        for s in subviews {
            let sz = s.sizeThatFits(.unspecified)
            if x > 0, x + sz.width > bounds.width { x = 0; y += rowH + spacing; rowH = 0 }
            s.place(at: CGPoint(x: bounds.minX + x, y: bounds.minY + y), proposal: ProposedViewSize(sz))
            x += sz.width + spacing
            rowH = max(rowH, sz.height)
        }
    }
}
